src/appleseed/renderer/kernel/shading/shadingpoint.cpp

```diff
@@ -358,6 +377,80 @@ void ShadingPoint::compute_partial_derivatives() const
 
         m_dpdu = normalize(Vector3d(tangent));
         m_dpdv = normalize(cross(sn, m_dpdu));
+        m_dndu = m_dndv = Vector3d(0.0);
+    }
+}
+
+void ShadingPoint::compute_screen_space_partial_derivatives() const
+{
+    const ShadingRay& ray = get_ray();
+
+    if (ray.m_has_differentials)
+    {
+        const Vector3d& n = get_shading_normal();
+        const Vector3d& p = get_point();
+        double t = intersect(ray.m_rx, p, n);
+        const Vector3d px = ray.m_rx.point_at(t);
+        m_dpdx = px - p;
+
+        t = intersect(ray.m_ry, p, n);
+        const Vector3d py = ray.m_ry.point_at(t);
+        m_dpdy = py - p;
+
+        size_t axes[2];
+        switch (max_abs_index(n))
```

Comment:
I'm sure we can find some clever alternative to that switch (which will likely be implemented as a jump table by the compiler). Here's a suggestion: http://codepad.org/sOvEIvs8

    const size_t max_index = max_abs_index(n);
    axes[0] = max_index == 0;
    axes[1] = 2 - (max_index >> 1);